server/src/main/java/org/apache/iotdb/db/mpp/execution/operator/process/DeviceViewIntoOperator.java

```diff
@@ -0,0 +1,140 @@
+import org.apache.iotdb.tsfile.utils.Binary;
+import org.apache.iotdb.tsfile.utils.Pair;
+
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.stream.Collectors;
+
+public class DeviceViewIntoOperator extends AbstractIntoOperator {
+
+  private final Map<String, Map<PartialPath, Map<String, InputLocation>>>
+      deviceToTargetPathSourceInputLocationMap;
+  private final Map<String, Map<PartialPath, Map<String, TSDataType>>>
+      deviceToTargetPathDataTypeMap;
+  private final Map<String, Boolean> targetDeviceToAlignedMap;
+  private final Map<String, List<Pair<String, PartialPath>>> deviceToSourceTargetPathPairListMap;
+
+  private String currentDevice;
+
+  private final TsBlockBuilder resultTsBlockBuilder;
+
+  public DeviceViewIntoOperator(
+      OperatorContext operatorContext,
+      Operator child,
+      Map<String, Map<PartialPath, Map<String, InputLocation>>>
+          deviceToTargetPathSourceInputLocationMap,
+      Map<String, Map<PartialPath, Map<String, TSDataType>>> deviceToTargetPathDataTypeMap,
+      Map<String, Boolean> targetDeviceToAlignedMap,
+      Map<String, List<Pair<String, PartialPath>>> deviceToSourceTargetPathPairListMap,
+      Map<String, InputLocation> sourceColumnToInputLocationMap) {
+    super(operatorContext, child, null, sourceColumnToInputLocationMap);
+    this.deviceToTargetPathSourceInputLocationMap = deviceToTargetPathSourceInputLocationMap;
+    this.deviceToTargetPathDataTypeMap = deviceToTargetPathDataTypeMap;
+    this.targetDeviceToAlignedMap = targetDeviceToAlignedMap;
+    this.deviceToSourceTargetPathPairListMap = deviceToSourceTargetPathPairListMap;
+
+    List<TSDataType> outputDataTypes =
+        ColumnHeaderConstant.selectIntoAlignByDeviceColumnHeaders.stream()
+            .map(ColumnHeader::getColumnType)
+            .collect(Collectors.toList());
+    this.resultTsBlockBuilder = new TsBlockBuilder(outputDataTypes);
+  }
+
+  @Override
+  public TsBlock next() throws IntoProcessException {
+    TsBlock inputTsBlock = child.next();
+    if (inputTsBlock != null) {
+      String device = String.valueOf(inputTsBlock.getValueColumns()[0].getBinary(0));
+      if (!Objects.equals(device, currentDevice)) {
+        insertMultiTabletsInternally(false);
+        updateResultTsBlock();
+
+        insertTabletStatementGenerators = constructInsertTabletStatementGeneratorsByDevice(device);
+        currentDevice = device;
+      }
+      int readIndex = 0;
+      while (readIndex < inputTsBlock.getPositionCount()) {
+        int lastReadIndex = readIndex;
+        for (IntoOperator.InsertTabletStatementGenerator generator :
+            insertTabletStatementGenerators) {
+          lastReadIndex =
+              Math.max(lastReadIndex, generator.processTsBlock(inputTsBlock, readIndex));
+        }
+        readIndex = lastReadIndex;
+        insertMultiTabletsInternally(true);
+      }
+    }
+
+    if (child.hasNext()) {
```

Comment:
It is hard to understand here, in the end of the next() method we check child.hasNext()? I think check hasNext() should be called outside and when hasNext()=true, then do next().

Reply:
fixed